# Домашнее задание 2 (5 баллов).

Все задания ниже имеют равный вес (5/16).

In [1]:
import pandas as pd

#### Описание данных

В папке Data находится информация о студентах. Всего 10 групп студентов. Файлы делятся на две категории:
    * Students_info_i - информация о студентах из группы i
    * Students_marks_i - оценки студентов из группы i за экзамены

### Одно из важных достоинств pandas $-$ это удобные методы реляционного взаимодействия с данными, аналогичные, например, возможностям SQL для слияния и конкатенации таблиц: merge, join, concat. Наличие готовых методов позволяет не реализовывать самостоятельно поэлементную обработку данных и оперировать сразу целыми таблицами данных.

Подробно об этих методах посмотрите тут: https://www.kaggle.com/residentmario/renaming-and-combining#Combining

#### Задание 1. Соберите всю информацию о студентах в одну таблицу df. В получившейся таблице должна быть информация и оценки всех студентов из всех групп. Напечатайте несколько строк таблицы для демонстрации результата.¶

In [2]:
!wget "https://github.com/hse-ds/iad-intro-ds/raw/master/2023/homeworks/Data.zip"
!unzip Data.zip

--2023-08-28 06:07:16--  https://github.com/hse-ds/iad-intro-ds/raw/master/2023/homeworks/Data.zip
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/hse-ds/iad-intro-ds/master/2023/homeworks/Data.zip [following]
--2023-08-28 06:07:17--  https://raw.githubusercontent.com/hse-ds/iad-intro-ds/master/2023/homeworks/Data.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 16727 (16K) [application/zip]
Saving to: ‘Data.zip’

Data.zip            100%[===================>]  16.33K  --.-KB/s    in 0s      

2023-08-28 06:07:17 (51.0 MB/s) - ‘Data.zip’ saved [16727/16727]

Archive:  Data.zip
   creating: Da

In [3]:
df = pd.read_csv("/content/Data/Students_info_0.csv").merge(pd.read_csv("/content/Data/Students_marks_0.csv"), "right", on="index")
for i in range(1,10):
  df = df.merge(pd.read_csv(f"/content/Data/Students_info_{i}.csv").merge(pd.read_csv(f"/content/Data/Students_marks_{i}.csv"), "right", on="index"), "outer")
df

index  gender race/ethnicity parental level of education         lunch  \
0        0  female        group B           bachelor's degree      standard   
1        1  female        group C                some college      standard   
2        2  female        group B             master's degree      standard   
3        3    male        group A          associate's degree  free/reduced   
4        4    male        group C                some college      standard   
..     ...     ...            ...                         ...           ...   
995    995  female        group E             master's degree      standard   
996    996    male        group C                 high school  free/reduced   
997    997  female        group C                 high school  free/reduced   
998    998  female        group D                some college      standard   
999    999  female        group D                some college  free/reduced   

    test preparation course    group  math score  reading score  writing score  
0                      none   group1          72             72             74  
1                 completed   group1          69             90             88  
2                      none   group1          90             95             93  
3                      none   group1          47             57             44  
4                      none   group1          76             78             75  
..                      ...      ...         ...            ...            ...  
995               completed  group10          88             99             95  
996                    none  group10          62             55             55  
997               completed  group10          59             71             65  
998               completed  group10          68             78             77  
999                    none  group10          77             86             86  

[1000 rows x 10 columns]

#### Задание 2. Удалите столбец index у полученной таблицы. Напечатайте первые 10 строк таблицы.

In [4]:
df.drop(columns = "index", inplace = True)
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75  
5                    none  group1          71             83             78  
6               completed  group1          88             95             92  
7                    none  group1          40             43             39  
8               completed  group1          64             64             67  
9                    none  group1          38             60             50

#### Задание 3. Выведите на экран размеры полученной таблицы

In [5]:
df.shape

(1000, 9)

#### Задание 4. Выведите на экран статистические характеристики числовых столбцов таблицы (минимум, максимум, среднее значение, стандартное отклонение)

In [6]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Задание 5. Проверьте, есть ли в таблице пропущенные значения

In [7]:
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
group                          0
math score                     0
reading score                  0
writing score                  0
dtype: int64

#### Задание 6. Выведите на экран средние баллы студентов по каждому предмету (math, reading, writing)

In [8]:
print("Средние баллы студентов по каждому предмету:\n", df[["math score", "reading score", "writing score"]].mean())


Средние баллы студентов по каждому предмету:
 math score       66.089
reading score    69.169
writing score    68.054
dtype: float64


**Задание 7. Как зависят оценки от того, проходил ли студент курс для подготовки к сдаче экзамена (test preparation course)? Выведите на экран для каждого предмета в отдельности средний балл студентов, проходивших курс для подготовки к экзамену и не проходивших курс.**

In [9]:
print("Средние баллы студентов, которые не проходили курс подготовки: \n",
df[df["test preparation course"]=="none"][["math score", "reading score", "writing score"]].mean(), "\n")

print("Средние баллы студентов, которые проходили курс подготовки: \n",
df[df["test preparation course"]=="completed"][["math score", "reading score", "writing score"]].mean())

Средние баллы студентов, которые не проходили курс подготовки: 
 math score       64.077882
reading score    66.534268
writing score    64.504673
dtype: float64 

Средние баллы студентов, которые проходили курс подготовки: 
 math score       69.695531
reading score    73.893855
writing score    74.418994
dtype: float64


**Задание 8. Выведите на экран все различные значения из столбца lunch.**

In [10]:
set(df.lunch)

{'free/reduced', 'standard'}

**Задание 9. Переименуйте колонку "parental level of education" в "education", а "test preparation course" в "test preparation" с помощью метода pandas rename**
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html

In [11]:
df.rename(columns = {"parental level of education" : "education", "test preparation course" :"test preparation"} , inplace = True)
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score  
0     group1          72             72             74  
1     group1          69             90             88  
2     group1          90             95             93  
3     group1          47             57             44  
4     group1          76             78             75  
..       ...         ...            ...            ...  
995  group10          88             99             95  
996  group10          62             55             55  
997  group10          59             71             65  
998  group10          68             78             77  
999  group10          77             86             86  

[1000 rows x 9 columns]

**Зафиксируем минимальный балл для сдачи экзамена**

In [12]:
passmark = 50

**Задание 10. Ответьте на вопросы:**
    * Какая доля студентов сдала экзамен по математике (passmark >= 50)?
    * Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?
    * Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике?

* Какая доля студентов сдала экзамен по математике (passmark >= 50)?

In [13]:
df[df["math score"] >=50]["math score"].count() / df["math score"].count()

0.865

**Ответ:** **86.5%** сдали экзамен по математике

* Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?

In [14]:
df[(df["test preparation"] == "completed") & (df["math score"] >=50)]["math score"].count() / df[(df["test preparation"] == "completed")]["math score"].count()

0.9217877094972067

**Ответ:** Приблизительно **92.179%** студентов, проходивших курс подготовки к экзамену, сдали экзамен по математике



* Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике?

In [15]:
df[(df.gender == "female") & (df["test preparation"] == "none") & (df["math score"] <50)]["math score"].count() /  df[(df.gender == "female") & (df["test preparation"] == "none")]["math score"].count()

0.20958083832335328

**Ответ:** Приблизительно **20.958%** женщин, не проходивших курс подготовки к экзамену, не сдали экзамен по математике

**Задание 11. С помощью groupby выполните задания ниже. Также выведите время выполнения каждого из заданий.**
    * Для каждой этнической группы выведите средний балл за экзамен по чтению
    * Для каждого уровня образования выведите минимальный балл за экзамен по письму

In [16]:
%%time
df.groupby("race/ethnicity").mean(numeric_only = True)["reading score"]

CPU times: user 3.95 ms, sys: 0 ns, total: 3.95 ms
Wall time: 13.5 ms


race/ethnicity
group A    64.674157
group B    67.352632
group C    69.103448
group D    70.030534
group E    73.028571
Name: reading score, dtype: float64

In [17]:
%%time
df.groupby("education").min(numeric_only = True)["writing score"]

CPU times: user 3.92 ms, sys: 2 µs, total: 3.92 ms
Wall time: 6.71 ms


education
associate's degree    35
bachelor's degree     38
high school           15
master's degree       46
some college          19
some high school      10
Name: writing score, dtype: int64

**Задание 12. Выполните задание 11 с помощью циклов. Сравните время выполнения.**

In [18]:
%%time
counter_by_group = dict(zip(list(set(df["race/ethnicity"])),[0,0,0,0,0]))
reading_sum_by_group = dict(zip(list(set(df["race/ethnicity"])),[0,0,0,0,0]))
writing_list_by_group = dict(zip(list(set(df["education"])),[[],[],[],[],[],[]]))
for i in range(len(df)):
  row = df[["race/ethnicity", "education", "reading score", "writing score"]].iloc[i,:]
  counter_by_group[row[0]] += 1
  reading_sum_by_group[row[0]] += row[2]
  writing_list_by_group[row[1]].append(row[3])
reading_mean_by_group = {}
writing_min_by_group = {}
for key in reading_sum_by_group.keys():
  reading_mean_by_group[key] = reading_sum_by_group[key] / counter_by_group[key]
for key in writing_list_by_group.keys():
    writing_min_by_group[key] = min(writing_list_by_group[key])
print(pd.DataFrame([reading_mean_by_group]).T.sort_index(),"\n",  pd.DataFrame([writing_min_by_group]).T.sort_index())

                 0
group A  64.674157
group B  67.352632
group C  69.103448
group D  70.030534
group E  73.028571 
                      0
associate's degree  35
bachelor's degree   38
high school         15
master's degree     46
some college        19
some high school    10
CPU times: user 1.15 s, sys: 16.7 ms, total: 1.16 s
Wall time: 1.26 s


**Задание 13. Выведите на экран средние баллы студентов по каждому предмету в зависимости от пола и уровня образования. То есть должно получиться количество групп, равных 2 * (число уровней образования), и для каждой такой группы выыведите средний балл по каждому из предметов.**

Это можно сделать с помощью сводных таблиц (pivot_table):

https://www.kaggle.com/kamilpolak/tutorial-how-to-use-pivot-table-in-pandas

In [19]:
pd.pivot_table(df, index = ['gender', 'education'],
               values = df[["math score", "reading score", "writing score"]])

math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000

#### Задание 14. Сколько студентов успешно сдали экзамен по математике?

Создайте новый столбец в таблице df под названием Math_PassStatus и запишите в него F, если студент не сдал экзамен по математике (балл за экзамен < passmark), и P иначе.

Посчитайте количество студентов, сдавших и не сдавших экзамен по математике.

Сделайте аналогичные шаги для экзаменов по чтению и письму.

In [20]:
def pass_or_fail(score):
  return "F" if score < passmark else "P"
df["Math_PassStatus"] = df["math score"].map(pass_or_fail)
df["Reading_PassStatus"] = df["reading score"].map(pass_or_fail)
df["Writing_PassStatus"] = df["writing score"].map(pass_or_fail)
print("Сдало экзамен по математике: ", df["Math_PassStatus"].value_counts()[0],
      "\nНе сдало экзамен по математике:", df["Math_PassStatus"].value_counts()[1],
      "\nСдало экзамен по reading:",df["Reading_PassStatus"].value_counts()[0],
      "\nНе сдало экзамен по reading:",df["Reading_PassStatus"].value_counts()[1],
      "\nСдало экзамен по writing:",df["Writing_PassStatus"].value_counts()[0],
      "\nНе сдало экзамен по writing:", df["Writing_PassStatus"].value_counts()[1])

Сдало экзамен по математике:  865 
Не сдало экзамен по математике: 135 
Сдало экзамен по reading: 910 
Не сдало экзамен по reading: 90 
Сдало экзамен по writing: 886 
Не сдало экзамен по writing: 114


#### Задание 15. Сколько студентов успешно сдали все экзамены?

Создайте столбец OverAll_PassStatus и запишите в него для каждого студента 'F', если студент не сдал хотя бы один из трех экзаменов, а иначе 'P'.

Посчитайте количество студентов, которые сдали все экзамены.

In [21]:
def pass_all_check(data):
  return "P" if (data["Math_PassStatus"] == "P") & (data["Reading_PassStatus"] == "P") & (data["Writing_PassStatus"] == "P") else "F"
df["OverAll_PassStatus"] = df.apply(pass_all_check,axis=1)
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus Writing_PassStatus OverAll_PassStatus  
0                    P                  P                  P  
1                    P                  P                  P  
2                    P                  P                  P  
3                    P                  F                  F  
4                    P                  P                  P  
..                 ...                ...                ...  
995                  P                  P                  P  
996                  P                  P                  P  
997                  P                  P                  P  
998                  P                  P                  P  
999                  P                  P                  P  

[1000 rows x 13 columns]

In [22]:
df[df["OverAll_PassStatus"] == "P"].shape[0]

812

Количество студентов, успешно сдавших все экзамены: **812**


#### Задание 16. Переведем баллы в оценки

### Система перевода баллов в оценки
####    больше 90 = A
####      80-90 = B
####      70-80 = C
####      60-70 = D
####      50-60 = E
####    меньше 50 = F (Fail)

Создайте вспомогательную функцию, которая будет по среднему баллу за три экзамена выставлять оценку студенту по данным выше критериям.

Создайте столбец Grade и запишите в него оценку каждого студента.

Выведите количество студентов, получивших каждую из оценок.

**В случае, если средний балл попадает на границу между оценками (т.е. равен ровно 60, 70 или 80 баллов), вы можете интерпретировать условие на своё усмотрение (т.е. можете поставить за 60 баллов оценку D, а можете - E).**

In [23]:
def GetGrade(average_mark):
    if average_mark >= 90:
        return 'A'
    elif average_mark >= 80:
        return 'B'
    elif average_mark >= 70:
        return 'C'
    elif average_mark >= 60:
        return 'D'
    elif average_mark >= 50:
        return 'E'
    else:
        return 'F'
df["Grade"] = ((df["math score"] + df["reading score"] + df["writing score"])/3).apply(GetGrade)
df["Grade"].value_counts(sort=True, ascending = True)

A     52
F    103
B    146
E    182
D    256
C    261
Name: Grade, dtype: int64